src/data/configs/cc_high_quality_20bn.yaml

```diff
@@ -0,0 +1,38 @@
+sources:
+  - name: cc_en_middle
+    source_coefficient: 0  # Remove all middle quality CC
+  - name: falcon
+    source_coefficient: 1
+  - name: cc_en_tail
+    source_coefficient: 0  # Remove all low quality CC
+  - name: cc_en_head
+    source_coefficient: 1  # TODO: Do we want to upsample cc_en_head to make sure CC fraction is same as with stratified?
```

Comment:
Agree with @orionw again, upsample, but the logic should definitely be `sample 3x the size from the larger amount of data`

Reply:
Here's the relevant code if helpful

```
sample_coefficients = {x["name"]: x["source_coefficient"] for x in config["sources"]}
# Estimate how many tokens we'd have without adjusting sample_coefficients
unadjusted_target_tokens = sum([v * src_cts[k] for k, v in sample_coefficients.items()])
# Adjust sample_coefficients uniformly to ensure desired target tokens
adjustment_ratio = target_tokens / unadjusted_target_tokens
sample_coefficients = {k: adjustment_ratio * v for k, v in sample_coefficients.items()}
```

With a toy example:

```
Source A: 100 tokens
Source B: 1000 tokens
```

If the config file specifies coefficients of `A: `5 and B: 1 and target tokens `300`.

I first compute the expected tokens after applying the scaling coefficients:

`5 * 100 + 1000 = 1500`

And then sampling A examples with probability

`5 * 300 / 1500 = 1.0`. `1.0 of 100 = 100 tokens`

and B with probability 

`1 * 300 / 1500 = 0.2`. `0.2 of 1000 = 200 tokens`

This gives us an expected token count of `300`. 

If there's a better way, feel free to share!

Reply:
The logic for how many tokens to gather seems solid, thanks!  I think that answers my question because it looks like you calculate the number of tokens to sample first, then sample.

Just to confirm though: the initial Source A would be representative of all the tokens in `dolma_20_percent_sample` correct? 

So when we have a coefficient of 5, we sample from all the possible data for that source in `dolma_20_percent_sample` rather than taking a 1x coefficient sample and then 5x upsampling it.  If that makes sense? So in the example, I think it would be the difference between sampling 100 tokens from 100 (e.g. taking all of them) vs sampling 20 tokens (since that's the 1x stratified sample representative amount) and then repeating them 5x.